crates/telio-starcast/src/starcast_peer.rs

```diff
@@ -0,0 +1,398 @@
+        // Set IPv4 header fields.
+        ipv4_packet.set_version(4);
+        ipv4_packet.set_header_length(5);
+        ipv4_packet.set_total_length(20 + 8 + payload.len() as u16);
+        ipv4_packet.set_ttl(64);
+        ipv4_packet.set_next_level_protocol(IpNextHeaderProtocols::Udp);
+        ipv4_packet.set_source(source_ip);
+        ipv4_packet.set_destination(destination_ip);
+
+        // Create a buffer for the UDP packet.
+        let mut udp_packet_buf = vec![0u8; 8 + payload.len()]; // UDP header + payload.
+        let mut udp_packet = MutableUdpPacket::new(&mut udp_packet_buf).unwrap();
+
+        // Set UDP header fields.
+        udp_packet.set_source(source_port);
+        udp_packet.set_destination(destination_port);
+        udp_packet.set_length(8 + payload.len() as u16);
+
+        // For simplicity, checksum is set to zero. In real applications, you should compute it.
+        udp_packet.set_checksum(0);
+
+        // Set the payload.
+        udp_packet.set_payload(payload);
+
+        // Insert UDP packet as IPV4 packet payload.
+        ipv4_packet.set_payload(udp_packet.packet());
+
+        ipv4_packet.packet().to_vec()
+    }
+
+    // This test connects two virtual peers by their sockets and tests if a packet
+    // can make it through between them.
+    #[tokio::test(flavor = "multi_thread")]
+    async fn test_two_virtual_peers_intercommunication() {
+        let (chan_a, chan_a_internal) = Chan::pipe();
+        let (mut chan_b, chan_b_internal) = Chan::pipe();
+        let starcast_allowed_ips = vec![
+            IpNetwork::V4(Ipv4Network::new(Ipv4Addr::new(224, 0, 0, 0), 4).unwrap()),
+            IpNetwork::V4(Ipv4Network::new(Ipv4Addr::new(100, 64, 0, 5), 32).unwrap()),
+            IpAddr::V6(Ipv6Addr::new(0xff02, 0, 0, 0, 0, 0, 0, 0xfb)).into(),
+        ];
+        let starcast_vpeer_a = StarcastPeer::start(chan_a_internal, starcast_allowed_ips.clone())
+            .await
+            .unwrap();
+        let starcast_vpeer_b = StarcastPeer::start(chan_b_internal, starcast_allowed_ips)
+            .await
+            .unwrap();
+
+        let a_peer = starcast_vpeer_a.get_peer().await.unwrap();
+        let b_peer = starcast_vpeer_b.get_peer().await.unwrap();
+
+        let peer_a_port = a_peer.endpoint.unwrap().port();
+        let peer_b_port = b_peer.endpoint.unwrap().port();
+
+        let config_a = Config {
+            public_key: b_peer.public_key,
+            wg_port: peer_b_port,
+        };
+        let config_b = Config {
+            public_key: a_peer.public_key,
+            wg_port: peer_a_port,
+        };
+
+        let udp_packet = create_udp_packet(peer_a_port, peer_b_port);
+
+        starcast_vpeer_a.configure(config_a).await.unwrap();
+        starcast_vpeer_b.configure(config_b).await.unwrap();
+
+        chan_a.tx.send(udp_packet.clone()).await.unwrap();
+        let received_bytes = await_timeout!(chan_b.rx.recv()).unwrap();
+
+        assert!(udp_packet == received_bytes)
```

Comment:
nit: assert_eq! would be better here - it will print both values when they differ.